File: ios/App/URB/Core/UrbResponder.swift
import Foundation

final class UrbResponder {
    private let id: String
    private var completed = false
    private let lock = NSLock()
    private let onComplete: ([String: Any]) -> Void

    init(id: String, onComplete: @escaping ([String: Any]) -> Void) {
        self.id = id
        self.onComplete = onComplete
    }

    func success(_ result: Any = [String: Any]()) {
        let response: [String: Any]
        lock.lock()
        if completed {
            lock.unlock()
            return
        }
        completed = true
        response = ["id": id, "ok": true, "result": result]
        lock.unlock()
        onComplete(response)
    }

    func error(code: String, message: String) {
        let response: [String: Any]
        lock.lock()
        if completed {
            lock.unlock()
            return
        }
        completed = true
        response = [
            "id": id,
            "ok": false,
            "error": ["code": code, "message": message]
        ]
        lock.unlock()
        onComplete(response)
    }
}

File: android/app/src/main/java/com/example/mywebview/urb/core/UrbResponder.kt
package com.example.mywebview.urb.core

import org.json.JSONObject
import java.util.concurrent.atomic.AtomicBoolean

class UrbResponder(
    private val id: String,
    private val onComplete: (String, JSONObject) -> Unit,
) {
    private val completed = AtomicBoolean(false)

    fun success(result: Any = JSONObject()) {
        complete(
            JSONObject()
                .put("id", id)
                .put("ok", true)
                .put("result", result),
        )
    }

    fun error(code: String, message: String) {
        complete(
            JSONObject()
                .put("id", id)
                .put("ok", false)
                .put(
                    "error",
                    JSONObject()
                        .put("code", code)
                        .put("message", message),
                ),
        )
    }

    private fun complete(response: JSONObject) {
        if (completed.compareAndSet(false, true)) {
            onComplete(id, response)
        }
    }
}
